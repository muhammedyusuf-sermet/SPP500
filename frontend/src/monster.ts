export enum Size {
	Tiny = "Tiny",
	Small = "Small",
	Medium = "Medium",
	Large = "Large",
	Huge = "Huge",
	Gargantuan = "Gargantuan"
}

export enum MonsterType {
	Beast = "Beast",
	Dragon = "Dragon",
	Monstrosity = "Monstrosity",
	Humanoid = "Humanoid",
	Fiend = "Fiend",
	Undead = "Undead",
	Giant = "Giant",
	Elemental = "Elemental",
	SwarmOfTinyBeasts = "SwarmOfTinyBeasts",
	Construct = "Construct",
	Celestial = "Celestial",
	Aberration = "Aberration",
	Fey = "Fey",
	Plant = "Plant",
	Ooze = "Ooze"
}

export enum MonsterRace {
	AnyRace = "AnyRace",
	Devil = "Devil",
	Demon = "Demon",
	Human = "Human",
	Shapechanger = "Shapechanger",
	Goblinoid = "Goblinoid",
	Titan = "Titan",
	Gnoll = "Gnoll",
	Gnome = "Gnome",
	Dwarf = "Dwarf",
	Elf = "Elf",
	Orc = "Orc",
	Kobold = "Kobold",
	Lizardfolk = "Lizardfolk",
	Merfolk = "Merfolk",
	Sahuagin = "Sahuagin",
	Grimlock = "Grimlock"
}

export enum Alignment {
	Unaligned = "Unaligned",
	AnyAlignment = "AnyAlignment",
	AnyNonGoodAlignment = "AnyNonGoodAlignment",
	AnyNonEvilAlignment = "AnyNonEvilAlignment",
	AnyNonLawfulAlignment = "AnyNonLawfulAlignment",
	AnyNonChaoticAlignment = "AnyNonChaoticAlignment",
	AnyGoodAlignment = "AnyGoodAlignment",
	AnyEvilAlignment = "AnyEvilAlignment",
	AnyLawfulAlignment = "AnyLawfulAlignment",
	AnyChaoticAlignment = "AnyChaoticAlignment",
	AnyNeutralAlignment = "AnyNeutralAlignment",
	LawfulGood = "LawfulGood",
	LawfulNeutral = "LawfulNeutral",
	LawfulEvil = "LawfulEvil",
	NeutralGood = "NeutralGood",
	Neutral = "Neutral",
	NeutralEvil = "NeutralEvil",
	ChaoticGood = "ChaoticGood",
	ChaoticNeutral = "ChaoticNeutral",
	ChaoticEvil = "ChaoticEvil"
}

export enum Environment {
	Arctic = "Arctic",
	Coastal = "Coastal",
	Desert = "Desert",
	Forest = "Forest",
	Grassland = "Grassland",
	Hill = "Hill",
	Mountain = "Mountain",
	Swamp = "Swamp",
	Underdark = "Underdark",
	Underwater = "Underwater",
	Urban = "Urban"
}

export enum MonsterAction {
	SpecialAbility = "SpecialAbility",
	Action = "Action",
	LegendaryAction = "LegendaryAction"
}

export interface IMonsterAbilities {
	name: string,
	description: string,
}

export enum TypeOfAction {
	Ranged = "Ranged",
	Melee = "Melee",
}

export interface IMonsterActions {
	name: string,
	description: string,
	usageCount: number,
	legendaryAction: boolean,
	typeOfAction?: TypeOfAction,
	toHit: number,
	reach: number,
	targets: number,
	averagedDamage: number,
	damageDice: number,
	damageConstant: number,

}

export interface IMonsterState {
	Id?: number,
	Name: string,
	Size?: Size | string;
	Type?: MonsterType | string;
	Race?: MonsterRace | string;
	Environment?: Environment | string;
	Alignment?: Alignment | string;

	ArmorClass?: number;
	HitPoints?: number;
	HitPointDistribution?: string;

	Speed?: string;
	Languages?: string;

	DamageVulnerabilities?: string;
	DamageResistances?: string;
	DamageImmunities?: string;
	ConditionImmunities?: string;

	ChallengeRating?: number;

	AbilityScores: IMonsterAbilityScoreState;
	Senses: SenseMap;
	Skills: SkillMap;
	SavingThrows: IMonsterSavingThrowState;
	//Actions: IActionState[];
}

export interface SkillMap {
	[skillName: string]: number|undefined;
}

export interface SenseMap {
	[senseName: string]: number|undefined;
}

export interface IMonsterAbilityScoreState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IMonsterSavingThrowState {
	[index:string]: number|undefined;
	Strength?: number;
	Dexterity?: number;
	Constitution?: number;
	Intelligence?: number;
	Wisdom?: number;
	Charisma?: number;
}

export interface IActionState {
	Name: string,
	Description: string,
	HitBonus?: number,
	Damage?: string,
	DamageBonus?: number,
	Type?: MonsterAction
}
/*export interface IMonster {
	name: string,
	type: MonsterType,
	alignment: MonsterAlignment,
	size: MonsterSize,
	race: MonsterRace,
	environment: MonsterEnvironment,
	resistance?: string,
	damageImmunity?: string,
	conditionImmunity?: string,
	vulnerability?: string,
	armorClass: number,
	hitPoints?: number,
	hitPointDice?: string,
	hitPointDiceAdd?: number,
	speedLand: number,
	speedSwim?: number,
	strStat: number,
	dexStat: number,
	conStat: number,
	intStat: number,
	wisStat: number,
	chaStat: number,
	strSavingThrow?: number,
	dexSavingThrow?: number,
	conSavingThrow?: number,
	intSavingThrow?: number,
	wisSavingThrow?: number,
	chaSavingThrow?: number,
	skillsAthletics: number,
	skillsAcrobatics?: number,
	skillsSleightOfHand?: number,
	skillsStealth?: number,
	skillsArcana?: number,
	skillsHistory?: number,
	skillsInvestigation?: number,
	skillsNature?: number,
	skillsReligion?: number,
	skillsAnimalHandling?: number,
	skillsInsight?: number,
	skillsMedicine?: number,
	skillsPerception?: number,
	skillsSurvival?: number,
	skillsDeception?: number,
	skillsIntimidation?: number,
	skillsPerformance?: number,
	skillsPersuasion?: number,
	sensesBlindsight?: number,
	sensesDarkvision?: number,
	sensesTremorsense?: number,
	sensesTruesight?: number,
	sensesPassivePerception?: number,
	sensesPassiveInvestigation?: number,
	sensesPassiveInsight?: number,
	languages?: string,
	challengeRating: number,
	experiencePoints: number,
	abilities?: IMonsterAbilities[],
	actions?: IMonsterActions[],
}*/