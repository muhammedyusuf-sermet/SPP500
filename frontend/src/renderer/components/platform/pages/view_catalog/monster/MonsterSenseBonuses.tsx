import * as React from 'react';
const Joi = require('joi');
import { ValidationError, ValidationOptions, JoiObject } from 'joi';

import 'bulma/css/bulma.css';

import { isDeepStrictEqual } from 'util';
import { ExpansionPanel, ExpansionPanelSummary, ExpansionPanelDetails, Typography, FormControl, InputLabel, Input, FormHelperText, Grid } from '@material-ui/core';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';

export interface IMonsterSenseBonusesProps {
	disabled?: boolean;
	// validation
	PayloadSchema: JoiObject;
	ValidationOptions: ValidationOptions;
	// initial values
	initial: {
		[senseName: string]: number | undefined;
	}
}

export interface IMonsterSenseBonusesState {
	[key: string]: number | string | undefined;
}

export class MonsterSenseBonuses extends React.Component<IMonsterSenseBonusesProps, IMonsterSenseBonusesState> {
	constructor(props: IMonsterSenseBonusesProps) {
		super(props);
		this.state = {
			...props.initial
		};
	}

	componentWillReceiveProps(nextProps: IMonsterSenseBonusesProps) {
		if (isDeepStrictEqual(this.props.initial, nextProps.initial) == false)
			this.setState({
				...nextProps.initial
			});
	}

	private keyNames: string[] = [
		'Blind', 'Blindsight', 'Darkvision', 'Tremorsense', 'Truesight',
		'Passive Perception', 'Passive Investigation', 'Passive Insight' ];

	stringToNumber = (toConvert : string) => {
		return isNaN(parseInt(toConvert)) ? undefined : parseInt(toConvert);
	}

	handleMonsterSenseChange = (event: React.ChangeEvent<HTMLInputElement>) => {
		const value = this.stringToNumber(event.target.value);
		const name = event.currentTarget.name;
		Joi.validate(
			{ [name]: value },
			Joi.reach(this.props.PayloadSchema, ['Senses']),
			this.props.ValidationOptions,
			(errors: ValidationError) => {
				this.setState({
					[name]: value,
					[name+'Error']: errors ? errors.details[0].message : undefined
				});
			});
		};

	render() {
		return (
			<ExpansionPanel defaultExpanded >
				<ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
					<Typography className='heading' >Sense Bonuses</Typography>
				</ExpansionPanelSummary>
				<ExpansionPanelDetails>
					<Grid container spacing={8} >
						{this.keyNames.map(senseName =>
							<Grid item xs={6} sm={4} key={senseName} >
								<FormControl className='formControl' fullWidth disabled={this.props.disabled} >
									<InputLabel htmlFor={senseName}>{senseName}</InputLabel>
									<Input
										id={senseName}
										type='number'
										value={this.state[senseName] != undefined ? this.state[senseName] : ''}
										name={senseName}
										onChange={this.handleMonsterSenseChange}
										aria-describedby={senseName+'-helper-text'} />
									<FormHelperText error id={senseName+'-helper-text'}>{this.state[senseName+'Error']}</FormHelperText>
								</FormControl>
							</Grid>
						)}
					</Grid>
				</ExpansionPanelDetails>
			</ExpansionPanel>
		);
	}
}