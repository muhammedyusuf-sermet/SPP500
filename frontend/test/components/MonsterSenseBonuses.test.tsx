import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import { IMonsterSenseBonusesProps, IMonsterSenseBonusesState, MonsterSenseBonuses } from '../../src/renderer/components/platform/pages/view_catalog/monster/MonsterSenseBonuses';
const Joi = require('joi');
import { ValidationOptions } from 'joi';

const payloadSchema = Joi.object({
	Senses: Joi.object().pattern(
		Joi.symbol().valid([
			"Blind", "Blindsight", "Darkvision", "Tremorsense", "Truesight",
			"Passive Perception", "Passive Investigation", "Passive Insight" ]),
		Joi.number().integer().greater(0).allow(0)
	).default({})
});
const validateOptions: ValidationOptions = {
	abortEarly: false,
	convert: true,
	allowUnknown: false,
	context: {

	}
};

////// Happy Path //////
describe('Monster SenseBonuses', () => {

	let monsterSenseBonusesInstance: ReactWrapper<IMonsterSenseBonusesProps, IMonsterSenseBonusesState, MonsterSenseBonuses>;

	describe('Provided Stat values', () => {

		beforeEach(() => {
			monsterSenseBonusesInstance =
				mount<MonsterSenseBonuses, IMonsterSenseBonusesProps, IMonsterSenseBonusesState>(
					<MonsterSenseBonuses
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{
							Blind: 30,
							Blindsight: 30,
							Darkvision: 10,
							Tremorsense: 15,
							Truesight: 60,
							"Passive Perception": 13,
							"Passive Investigation": 14,
							"Passive Insight": 16,
						}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterSenseBonusesInstance).toBeDefined();
		});

		it('should have the correct number for Blind', () => {
			expect(monsterSenseBonusesInstance.find('Input#Blind').props().value).toEqual(30);
		});

		it('should have the correct number for Blindsight', () => {
			expect(monsterSenseBonusesInstance.find('Input#Blindsight').props().value).toEqual(30);
		});

		it('should have the correct number for Darkvision', () => {
			expect(monsterSenseBonusesInstance.find('Input#Darkvision').props().value).toEqual(10);
		});

		it('should have the correct number for Tremorsense', () => {
			expect(monsterSenseBonusesInstance.find('Input#Tremorsense').props().value).toEqual(15);
		});

		it('should have the correct number for Truesight', () => {
			expect(monsterSenseBonusesInstance.find('Input#Truesight').props().value).toEqual(60);
		});

		it('should have the correct number for Passive Perception', () => {
			expect(monsterSenseBonusesInstance.find('Input[id="Passive Perception"]').props().value).toEqual(13);
		});

		it('should have the correct number for Passive Investigation', () => {
			expect(monsterSenseBonusesInstance.find('Input[id="Passive Investigation"]').props().value).toEqual(14);
		});

		it('should have the correct number for Passive Insight', () => {
			expect(monsterSenseBonusesInstance.find('Input[id="Passive Insight"]').props().value).toEqual(16);
		});
	});

	describe('Happy Path', () => {

		beforeEach(() => {
			monsterSenseBonusesInstance =
				mount<MonsterSenseBonuses, IMonsterSenseBonusesProps, IMonsterSenseBonusesState>(
					<MonsterSenseBonuses
						disabled={false}
						PayloadSchema={payloadSchema}
						ValidationOptions={validateOptions}
						initial={{}} />
				);
		})

		it('renders without crashing', () => {
			expect(monsterSenseBonusesInstance).toBeDefined();
		});

		it('should be able to update state', () => {
			monsterSenseBonusesInstance.find('Input#Blind').simulate('change', { target: { value: 9 } })
			monsterSenseBonusesInstance.find('Input#Blindsight').simulate('change', { target: { value: 10 } })
			monsterSenseBonusesInstance.find('Input#Darkvision').simulate('change', { target: { value: 8 } })
			monsterSenseBonusesInstance.find('Input#Tremorsense').simulate('change', { target: { value: 7 } })
			monsterSenseBonusesInstance.find('Input#Truesight').simulate('change', { target: { value: 7 } })
			monsterSenseBonusesInstance.find('Input[id="Passive Perception"]').simulate('change', { target: { value: 2 } })
			monsterSenseBonusesInstance.find('Input[id="Passive Investigation"]').simulate('change', { target: { value: 14 } })
			monsterSenseBonusesInstance.find('Input[id="Passive Insight"]').simulate('change', { target: { value: 30 } })
			expect(monsterSenseBonusesInstance.state()).toEqual({
				// data
				Blind: 9,
				Blindsight: 10,
				Darkvision: 8,
				Tremorsense: 7,
				Truesight: 7,
				"Passive Perception": 2,
				"Passive Investigation": 14,
				"Passive Insight": 30,
				// errors
				BlindError: undefined,
				BlindsightError: undefined,
				DarkvisionError: undefined,
				TremorsenseError: undefined,
				TruesightError: undefined,
				"Passive PerceptionError": undefined,
				"Passive InvestigationError": undefined,
				"Passive InsightError": undefined,
			});
		});
		// STATE CHANGE
		it('should change only Blind when Blind is changed', () => {
			monsterSenseBonusesInstance.find('Input#Blind').simulate('change', { target: { value: '9' } })
			expect(monsterSenseBonusesInstance.state().Blind).toEqual(9);
		})

		it('should change only Blindsight when Blindsight is changed', () => {
			monsterSenseBonusesInstance.find('Input#Blindsight').simulate('change', { target: { value: '10' } })
			expect(monsterSenseBonusesInstance.state().Blindsight).toEqual(10);
		})

		it('should change only Darkvision when Darkvision is changed', () => {
			monsterSenseBonusesInstance.find('Input#Darkvision').simulate('change', { target: { value: '0' } })
			expect(monsterSenseBonusesInstance.state().Darkvision).toEqual(0);
		})

		it('should change only Tremorsense when Tremorsense is changed', () => {
			monsterSenseBonusesInstance.find('Input#Tremorsense').simulate('change', { target: { value: '7' } })
			expect(monsterSenseBonusesInstance.state().Tremorsense).toEqual(7);
		})

		it('should change only Truesight when Truesight is changed', () => {
			monsterSenseBonusesInstance.find('Input#Truesight').simulate('change', { target: { value: '7' } })
			expect(monsterSenseBonusesInstance.state().Truesight).toEqual(7);
		})

		it('should change only Passive Perception when Passive Perception is changed', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Perception"]').simulate('change', { target: { value: '2' } })
			expect(monsterSenseBonusesInstance.state()['Passive Perception']).toEqual(2);
		})

		it('should change only Passive Investigation when Passive Investigation is changed', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Investigation"]').simulate('change', { target: { value: '14' } })
			expect(monsterSenseBonusesInstance.state()['Passive Investigation']).toEqual(14);
		})

		it('should change only Passive Insight when Passive Insight is changed', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Insight"]').simulate('change', { target: { value: '20' } })
			expect(monsterSenseBonusesInstance.state()['Passive Insight']).toEqual(20);
		})

		// DEFAULTS
		it('should change Blind back to undefined when input is empty', () => {
			monsterSenseBonusesInstance.find('Input#Blind').simulate('change', { target: { value: '' } })
			expect(monsterSenseBonusesInstance.state().Blind).toEqual(undefined);
		})

		it('should change Blindsight back to undefined when input is empty', () => {
			monsterSenseBonusesInstance.find('Input#Blindsight').simulate('change', { target: { value: '' } })
			expect(monsterSenseBonusesInstance.state().Blindsight).toEqual(undefined);
		})

		it('should change Darkvision back to undefined when input is empty', () => {
			monsterSenseBonusesInstance.find('Input#Darkvision').simulate('change', { target: { value: '' } })
			expect(monsterSenseBonusesInstance.state().Darkvision).toEqual(undefined);
		})

		it('should change Tremorsense back to undefined when input is empty', () => {
			monsterSenseBonusesInstance.find('Input#Tremorsense').simulate('change', { target: { value: '' } })
			expect(monsterSenseBonusesInstance.state().Tremorsense).toEqual(undefined);
		})

		it('should change Passive Perception back to undefined when input is empty', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Perception"]').simulate('change', { target: { value: '' } })
			expect(monsterSenseBonusesInstance.state()['Passive Perception']).toEqual(undefined);
		})

		it('should change Passive Investigation back to undefined when input is empty', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Investigation"]').simulate('change', { target: { value: '' } })
			expect(monsterSenseBonusesInstance.state()['Passive Investigation']).toEqual(undefined);
		})

		it('should change Passive Insight back to undefined when input is empty', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Insight"]').simulate('change', { target: { value: '' } })
			expect(monsterSenseBonusesInstance.state()['Passive Insight']).toEqual(undefined);
		})

		// VALIDATION
		it('should validate Blind when Blind is changed', () => {
			monsterSenseBonusesInstance.find('Input#Blind').simulate('change', { target: { value: '-12'  } })
			expect(monsterSenseBonusesInstance.state().Blind).toEqual(-12);
			expect(monsterSenseBonusesInstance.find('Help#Blind').text()).toEqual('\"Blind\" must be greater than 0');
		})

		it('should validate Blindsight when Blindsight is changed', () => {
			monsterSenseBonusesInstance.find('Input#Blindsight').simulate('change', { target: { value: '-102'  } })
			expect(monsterSenseBonusesInstance.state().Blindsight).toEqual(-102);
			expect(monsterSenseBonusesInstance.find('Help#Blindsight').text()).toEqual('\"Blindsight\" must be greater than 0');
		})

		it('should validate Darkvision when Darkvision is changed', () => {
			monsterSenseBonusesInstance.find('Input#Darkvision').simulate('change', { target: { value: '-1'  } })
			expect(monsterSenseBonusesInstance.state().Darkvision).toEqual(-1);
			expect(monsterSenseBonusesInstance.find('Help#Darkvision').text()).toEqual('\"Darkvision\" must be greater than 0');
		})

		it('should validate Tremorsense when Tremorsense is changed', () => {
			monsterSenseBonusesInstance.find('Input#Tremorsense').simulate('change', { target: { value: '-9'  } })
			expect(monsterSenseBonusesInstance.state().Tremorsense).toEqual(-9);
			expect(monsterSenseBonusesInstance.find('Help#Tremorsense').text()).toEqual('\"Tremorsense\" must be greater than 0');
		})

		it('should validate Truesight when Truesight is changed', () => {
			monsterSenseBonusesInstance.find('Input#Truesight').simulate('change', { target: { value: '-12'  } })
			expect(monsterSenseBonusesInstance.state().Truesight).toEqual(-12);
			expect(monsterSenseBonusesInstance.find('Help#Truesight').text()).toEqual('\"Truesight\" must be greater than 0');
		})

		it('should validate Passive Perception when Passive Perception is changed', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Perception"]').simulate('change', { target: { value: '-16'  } })
			expect(monsterSenseBonusesInstance.state()["Passive Perception"]).toEqual(-16);
			expect(monsterSenseBonusesInstance.find('Help[id="Passive Perception"]').text()).toEqual('\"Passive Perception\" must be greater than 0');
		})

		it('should validate Passive Investigation when Passive Investigation is changed', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Investigation"]').simulate('change', { target: { value: '-6'  } })
			expect(monsterSenseBonusesInstance.state()["Passive Investigation"]).toEqual(-6);
			expect(monsterSenseBonusesInstance.find('Help[id="Passive Investigation"]').text()).toEqual('\"Passive Investigation\" must be greater than 0');
		})

		it('should validate Passive Insight when Passive Insight is changed', () => {
			monsterSenseBonusesInstance.find('Input[id="Passive Insight"]').simulate('change', { target: { value: '-1'  } })
			expect(monsterSenseBonusesInstance.state()["Passive Insight"]).toEqual(-1);
			expect(monsterSenseBonusesInstance.find('Help[id="Passive Insight"]').text()).toEqual('\"Passive Insight\" must be greater than 0');
		})
	});
});