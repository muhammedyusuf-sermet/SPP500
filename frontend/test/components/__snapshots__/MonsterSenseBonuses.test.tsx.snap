// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monster SenseBonuses Small snapshot matches snapshot 1`] = `
<WithStyles(ExpansionPanel)
  CollapseProps={
    Object {
      "timeout": 100,
    }
  }
  defaultExpanded={true}
>
  <WithStyles(ExpansionPanelSummary)
    expandIcon={<pure(ExpandMoreIcon) />}
  >
    <WithStyles(Typography)
      className="heading"
    >
      Sense Bonuses
    </WithStyles(Typography)>
  </WithStyles(ExpansionPanelSummary)>
  <WithStyles(ExpansionPanelDetails)>
    <WithStyles(Grid)
      container={true}
      spacing={8}
    >
      <WithStyles(Grid)
        item={true}
        key="Blind"
        sm={4}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Blind"
          >
            Blind
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Blind-helper-text"
            id="Blind"
            name="Blind"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Blind-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="Blindsense is a lesser ability that lets the creature notice things it cannot see, but without the precision of blindsight, using nonvisual senses, such as acute smell or hearing."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About 
              Blind
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="Blindsight"
        sm={4}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Blindsight"
          >
            Blindsight
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Blindsight-helper-text"
            id="Blindsight"
            name="Blindsight"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Blindsight-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="Using nonvisual senses, such as sensitivity to vibrations, keen smell, acute hearing, or echolocation, a creature with blindsight maneuvers and fights as well as a sighted creature. Invisibility, darkness, and most kinds of concealment are irrelevant, though the creature must have line of effect to a creature or object to discern that creature or object."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About 
              Blindsight
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="Darkvision"
        sm={4}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Darkvision"
          >
            Darkvision
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Darkvision-helper-text"
            id="Darkvision"
            name="Darkvision"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Darkvision-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="A monster with Darkvision can see in the dark within a specific radius. The monster can see in dim light within the radius as if it were bright light, and in Darkness as if it were dim light. The monster can’t discern color in Darkness, only shades of gray."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About 
              Darkvision
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="Tremorsense"
        sm={4}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Tremorsense"
          >
            Tremorsense
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Tremorsense-helper-text"
            id="Tremorsense"
            name="Tremorsense"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Tremorsense-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="A monster with tremorsense can detect and pinpoint the Origin of vibrations within a specific radius, provided that the monster and the source of the vibrations are in contact with the same ground or substance. Tremorsense can’t be used to detect flying or incorporeal creatures."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About 
              Tremorsense
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="Truesight"
        sm={4}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Truesight"
          >
            Truesight
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Truesight-helper-text"
            id="Truesight"
            name="Truesight"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Truesight-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="A creature with truesight can, out to a specific range, see in normal and magical darkness, see invisible creatures and objects, automatically detect visual illusions and succeed on saving throws against them, and perceives the original form of a shapechanger or a creature that is transformed by magic. Furthermore, the creature can see into the Ethereal Plane."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About 
              Truesight
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="Passive Perception"
        sm={4}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Passive Perception"
          >
            Passive Perception
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Passive Perception-helper-text"
            id="Passive Perception"
            name="Passive Perception"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Passive Perception-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="This stat helps to handle whether monsters who are not actively searching for something will notice a hidden object or creature."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About 
              Passive Perception
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="Passive Investigation"
        sm={4}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Passive Investigation"
          >
            Passive Investigation
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Passive Investigation-helper-text"
            id="Passive Investigation"
            name="Passive Investigation"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Passive Investigation-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="This stat helps to handle whether monsters who are not actively trying to investigate something will notice a clue or fact about that thing."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About 
              Passive Investigation
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        key="Passive Insight"
        sm={4}
        xs={6}
      >
        <WithStyles(FormControl)
          className="formControl"
          disabled={false}
          fullWidth={true}
        >
          <WithStyles(WithFormControlContext(InputLabel))
            htmlFor="Passive Insight"
          >
            Passive Insight
          </WithStyles(WithFormControlContext(InputLabel))>
          <WithStyles(Input)
            aria-describedby="Passive Insight-helper-text"
            id="Passive Insight"
            name="Passive Insight"
            onChange={[Function]}
            type="number"
            value=""
          />
          <WithStyles(WithFormControlContext(FormHelperText))
            error={true}
            id="Passive Insight-helper-text"
          />
          <WithStyles(Tooltip)
            disableFocusListener={true}
            title="This stat helps to handle whether monsters who are not actively trying to detect deception will notice when someone is lying to or witholding truth from them."
          >
            <withHelpersModifiers(Help)
              className="tooltip"
            >
              Hover for More About 
              Passive Insight
            </withHelpersModifiers(Help)>
          </WithStyles(Tooltip)>
        </WithStyles(FormControl)>
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(ExpansionPanelDetails)>
</WithStyles(ExpansionPanel)>
`;
